### PR TITLE
Make default layer size 16-bit

### PR DESCRIPTION
**Pull request by drashna**
## Description

Set the layer state to be 16-bit by default.

I would rather set this to 8-bit, but until the configurator can handle that without issue, ... it supports 16 layers, so this would work in the meanwhile.

And this allows for smaller compiled sizes.

## Types of Changes

- [x] Core
- [x] Enhancement/optimization
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by KarlK90 — APPROVED** [review]
LGTM and is a sane default.

**Comment by precondition** [comment]
I can already hear the unrest that would come from the havoc caused by all the `biton32` in the wild no longer working. Other than that, I'm in favor of this.

**Comment by fauxpark** [comment]
That's why you're supposed to use `get_highest_layer()` instead ;)

**Comment by precondition** [comment]
Yes, indeed but I still see too many people on the QMK discord server posting code bits using `biton32` instead of `get_highest_layer` :(

**Review by precondition — APPROVED** [review]

**Comment by precondition** [comment]
Speaking of `biton32`, shouldn't the ~265 user files calling the `biton32` function in place of `get_highest_layer` get changed in this PR too? For example:
```
keyboards/bpiphany/kitten_paw/keymaps/ickerwx/keymap.c:135:    layer = biton32(layer_state);
```

**Comment by fauxpark** [comment]
We try to avoid changing user code en masse like that...

**Comment by drashna** [comment]
> Yes, indeed but I still see too many people on the QMK discord server posting code bits using `biton32` instead of `get_highest_layer` :(

That's because it in a lot of the default keymaps.  And people copy and paste code. 

But I've been flagging that in any new PRs, so more copy-paste code hasn't been incorrectly added.

**Comment by drashna** [comment]
> Speaking of `biton32`, shouldn't the ~265 user files calling the `biton32` function in place of `get_highest_layer` get changed in this PR too? For example:
> 
> ```
> keyboards/bpiphany/kitten_paw/keymaps/ickerwx/keymap.c:135:    layer = biton32(layer_state);
> ```

Also, some of the rational behind not changing all of the user files doing this (besides, "that's a lot of work"), is that using biton32 will work and "fail gracefully". So there is no harm in leaving it. But future examples and the default keymps *should* use the correct function.

**Comment by zvecr** [comment]
Also worth noting that the configurator currently supports more than 8 layers.

**Comment by drashna** [comment]
> Also worth noting that the configurator currently supports more than 8 layers. And some keyboards also set `DYNAMIC_KEYMAP_LAYER_COUNT` greater than 8.

The dynamic keymap layer count is already taken into account, but github doesn't show a lot of context.  Literally, right above: 

https://github.com/qmk/qmk_firmware/blob/c76cfc2d36ead86a5ec9e4d5061a28987b0bca18/quantum/action_layer.h#L24-L41

However, the configurator is a good point, and this PR should wait to be merged until there is something in place to handle that.

**Comment by noroadsleft** [comment]
Merge conflicts.

File: keyboards/helix/rev3_5rows/keymaps/five_rows/keymap.c [hunk 1]
```diff
@@ -362,7 +362,7 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
 
 int current_default_layer;
 
-uint32_t default_layer_state_set_user(uint32_t state) {
+layer_state_t default_layer_state_set_user(layer_state_t state) {
     current_default_layer = biton32(state);
```
Comment (fauxpark) [inline keyboards/helix/rev3_5rows/keymaps/five_rows/keymap.c:366]:
```suggestion
    current_default_layer = get_highest_layer(state);
```
?

File: keyboards/helix/rev2/keymaps/five_rows/keymap.c [hunk 2]
```diff
@@ -362,7 +362,7 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
 
 int current_default_layer;
 
-uint32_t default_layer_state_set_user(uint32_t state) {
+layer_state_t default_layer_state_set_user(layer_state_t state) {
     current_default_layer = biton32(state);
```
Comment (fauxpark) [inline keyboards/helix/rev2/keymaps/five_rows/keymap.c:366]:
```suggestion
    current_default_layer = get_highest_layer(state);
```

File: keyboards/helix/rev2/keymaps/edvorakjp/keymap.c [hunk 3]
```diff
@@ -5,7 +5,7 @@
 // keymaps definitions are moved to keymap_Xrows.c.
 
 #ifdef RGBLIGHT_ENABLE
-uint32_t layer_state_set_keymap(uint32_t state) {
+layer_state_t layer_state_set_keymap(layer_state_t state) {
     rgblight_mode_noeeprom(RGBLIGHT_MODE_STATIC_LIGHT);
     switch (biton32(state)) {
```
Comment (fauxpark) [inline keyboards/helix/rev2/keymaps/edvorakjp/keymap.c:10]:
```suggestion
    switch (get_highest_layer(state)) {
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by drashna** [comment]
> Also worth noting that the configurator currently supports more than 8 layers. And some keyboards also set `DYNAMIC_KEYMAP_LAYER_COUNT` greater than 8.

Default set to 16, which matches the max of what the configurator supports.  So, between that, and the "autodetection" based on the dynamic layers define, it should cover all the bases, for now.

File: quantum/action_layer.h [hunk 4]
```diff
@@ -64,7 +64,7 @@ typedef uint32_t layer_state_t;
 #    ifndef MAX_LAYER
 #        define MAX_LAYER 32
 #    endif
-#    define get_highest_layer(state) biton32(state)
+#    define get_highest_layer(state) get_highest_layer(state)
```
Comment (noroadsleft) [inline quantum/action_layer.h:67]: :thinking:

Comment (drashna) [reply quantum/action_layer.h:67]: blind find replace strikes again. 😅 
```suggestion
#    define get_highest_layer(state) biton32(state)
```

**Comment by fauxpark** [comment]
Merge conflicts :(

**Review by noroadsleft — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Comment by tzarc** [comment]
Seems to break the helix boards.

**Comment by tzarc** [comment]
Deferred to Q3 so helix can be fixed and verified.

**Comment by drashna** [comment]
> Deferred to Q3 so helix can be fixed and verified.

How was this breaking the helix boards?  

I do see one keymap that won't compile, but that's because it's doing some really weird stuff.  All the rest compile.  However, I don't have a board to test on.

**Comment by tzarc** [comment]
I don't remember, and at this stage I'm inclined to say "develop can break, we can fix over time if issues occur".

**Pull request merged**